### PR TITLE
[receiver/datadog] Adopt SharedComponent

### PR DESCRIPTION
**Pull request by boostchicken**
**Description:** Adopt SharedComponent, this moves DD Receiver creation limits into the proper component

**Link to tracking Issue:** <Issue number if applicable>

**Testing:** Existing unit tests

**Documentation:** <Describe the documentation added.>

### CODE REVIEW

File: receiver/datadogreceiver/receiver.go [hunk 1]
```diff
@@ -61,30 +56,21 @@ func newDataDogReceiver(config *Config, nextConsumer consumer.Traces, params rec
...
-			if err := ddr.server.ListenAndServe(); err != http.ErrServerClosed {
-				host.ReportFatalError(fmt.Errorf("error starting datadog receiver: %w", err))
-			}
-		}()
-	})
+	go func() {
+		ddmux := http.NewServeMux()
+		ddmux.HandleFunc("/v0.3/traces", ddr.handleTraces)
+		ddmux.HandleFunc("/v0.4/traces", ddr.handleTraces)
+		ddmux.HandleFunc("/v0.5/traces", ddr.handleTraces)
```
Comment (jpkrohling) [inline receiver/datadogreceiver/receiver.go:63]: Is this change here competing with the one adding support for 0.7?

File: receiver/datadogreceiver/factory.go [hunk 2]
```diff
@@ -46,11 +48,13 @@ func createDefaultConfig() component.Config {
...
 	rcfg := cfg.(*Config)
-	return newDataDogReceiver(rcfg, consumer, params)
+	r = receivers.GetOrAdd(cfg, func() component.Component {
+		dd, _ := newDataDogReceiver(rcfg, consumer, params)
+		return dd
+	})
+	return r, nil
 }
+
+var receivers = sharedcomponent.NewSharedComponents()
```
Comment (jpkrohling) [inline receiver/datadogreceiver/factory.go:60]: Is this pattern being used elsewhere? Is this new?

File: receiver/datadogreceiver/receiver.go (hunk 1 above)
Comment (boostchicken) [reply receiver/datadogreceiver/receiver.go:63]: A conflict is inevitable

File: receiver/datadogreceiver/factory.go (hunk 2 above)
Comment (boostchicken) [reply receiver/datadogreceiver/factory.go:60]: No its been around and a lot of receivers and exports use it

**Review by jpkrohling — APPROVED** [review]

**Review by TylerHelmuth — APPROVED** [review]

**Comment by TylerHelmuth** [comment]
@boostchicken please take a look at the merge conflicts

**Comment by boostchicken** [comment]
@jpkrohling fixed the conflicts

**Pull request merged**